### PR TITLE
Implement Brent's method for 1d optimisation

### PR DESCRIPTION
**Pull request by richfitz**
Another good derivative-free optimisation algorithm (which we'll likely use to polish the extremes in odin-js's batch runs), Brent's algorithm for 1d optimisation.

There are convergence guarantees that exist in 1d that don't in higher dimensions, so this is worth a slightly different interface. The algorithm here will outperform simplex on a 1d problem I'd expect, and should do quite well on our interpolated ODE traces.

### CODE REVIEW

File: src/brent.ts [hunk 1]
```diff
@@ -0,0 +1,341 @@
+import { checkResult, Result, TargetFn1 } from "./types";
+import { copysign, invert, protect, withDefault } from "./utils";
+
+/**
+ * Control for the Brent algorithm. Only some of these (or, indeed none of
+ * these) can be provided to {@link Brent} or {@link fitBrent}
+ */
+export interface BrentControlParam {
```
Comment (EmmaLRussell) [inline src/brent.ts:8]: For consistency, should these be in `control.ts`? (Or `SimplexControlParam` moved into `simplex.ts`?)

File: src/brent.ts [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+    public step() {
+        this._converged = this._step();
+        return this._converged;
+    }
+
+    /**
+     * Helper function to run the algorithm until converged. This is
+     * very basic and not really intended to be used - you should
+     * probably build logic around {@link Simplex.step} directly, or if
+     * you want a simple interface use the {@link fitSimplex} function.
```
Comment (EmmaLRussell) [inline src/brent.ts]:
```suggestion
     * probably build logic around {@link Brent.step} directly, or if
     * you want a simple interface use the {@link fitBrent} function.
```

**Review by EmmaLRussell — COMMENTED** [review]
> which we'll likely use to polish the extremes in odin-js's batch runs

So, once this is used in odin-js, might the result of `wodinFit` be a Brent rather than a Simplex? I guess Brent and Simplex could support a shared interface - for step(), result() etc

Sorry, my brain got a bit lost in the maths (the YouTube video helped though!), but I appreciated the extensive commenting which helped give some commentary to what's going on here!

File: src/brent.ts (hunk 1 above)
Comment (richfitz) [reply src/brent.ts:8]: good point, I'll rename simplex.ts -> simplex-control.ts and do similar for brent, or move the simplex control into simplex.ts

File: src/brent.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+ *
+ * We track a set of points on the domain [a, b, u, v, w, x], not all
+ * distinct.
+ *
+ * * Initially (`a`, `b`) is the domain of `target` where we will search
+ *   for the minimum.
+ * * `x` is the point visited with the lowest value of `target`
+ * * `w` is the second lowest visited value of `target`
+ * * `v` is the previous value of `w`
+ * * `fx`, `fw` and `fv` are the values of `target` evaluated at `x`, `w`
```
Comment (EmmaLRussell) [inline src/brent.ts]: Is this just explanatory, or was the intention to include fx etc in state?

Comment (richfitz) [reply src/brent.ts]: this was before a refactor - there used to be elements x, w, v, fx, fw, fv but I merged them into the `Point1` interface which simplified the bookkeeping. Will update the comment

**Comment by richfitz** [comment]
> So, once this is used in odin-js, might the result of wodinFit be a Brent rather than a Simplex? I guess Brent and Simplex could support a shared interface - for step(), result() etc

We can't actually use this for `wodinFit` because Brent's requires a bounded interval to search in (Simplex searches from a point anywhere in n-dimensional space). Where we'd use this is transparently in the `findExtremes()` function, via the one-shot `fitBrent` function

File: src/brent.ts (hunk 1 above)
Comment (richfitz) [reply src/brent.ts:8]: Done in f884538 which makes this a bit more churny (sorry Alex!)

**Review by hillalex — APPROVED** [review]

**Pull request merged**